blocks/monitoring/evaluators.py

```diff
@@ -186,33 +269,51 @@ def _compile(self):
             be out-sourced to `ComputationGraph` to deal with it.
 
         """
-        if self.buffer_.accumulation_updates:
+        inputs = []
+        outputs = []
+        updates = None
+        if self.theano_buffer.accumulation_updates:
             updates = OrderedDict()
-            updates.update(self.buffer_.accumulation_updates)
+            updates.update(self.theano_buffer.accumulation_updates)
             if self.updates:
                 updates.update(self.updates)
-            self._accumulate_fun = theano.function(
-                self.buffer_.inputs, [],
-                updates=updates)
+            inputs += self.theano_buffer.inputs
+        inputs += self.monitored_quantities_buffer.inputs
+        outputs = self.monitored_quantities_buffer.requires
+
+        if inputs != []:
+            unique_inputs = list(set(inputs))
+            self._accumulate_fun = theano.function(unique_inputs,
+                                                   outputs,
+                                                   updates=updates)
         else:
             self._accumulate_fun = None
 
     def initialize_aggregators(self):
-        self.buffer_.initialize_aggregators()
+        self.theano_buffer.initialize_aggregators()
+        self.monitored_quantities_buffer.initialize()
 
     def process_batch(self, batch):
         try:
-            batch = dict_subset(batch, self.buffer_.input_names)
+            input_names = self.theano_buffer.input_names + \
```

Comment:
I feel bad about this code, because you will have duplicates in `input_names`. Can you make `unique_inputs` from 285 an attribute, i.e. `self.unique_inputs`, and reuse it here?